File: Static-Deploy/build_scripts/torflux.c
// Copyright(C) 2024 NOXCIS [https://github.com/NOXCIS]
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <sys/stat.h>

#define LOG_DIR "./log"
#define LOG_FILE LOG_DIR "/tor_circuit_refresh_log.txt"
#define BUFFER_SIZE 1024
#define TOR_CONTROL_PORT_1 9051
#define TOR_CONTROL_PORT_2 9054

// Function prototypes
void log_message(const char *message, int add_newlines, int to_console);
int send_signal(int port, const char *password, char *status_buffer);
int circuits_are_different(const char *old_status, const char *new_status);
void get_timestamp(char *buffer, size_t size);

// Function to log messages
void log_message(const char *message, int add_newlines, int to_console) {
    FILE *log_f;
    char timestamp[64];
    get_timestamp(timestamp, sizeof(timestamp));

    // Format the log message with timestamp
    char log_message[BUFFER_SIZE];
    snprintf(log_message, BUFFER_SIZE, "[%s] %s", timestamp, message);

    // Print to console if requested
    if (to_console) {
        printf("%s\n", log_message);
    }

    // Ensure log directory exists
    mkdir(LOG_DIR, 0755);

    // Write to the log file
    log_f = fopen(LOG_FILE, "a");
    if (log_f) {
        fprintf(log_f, "%s\n", log_message);
        if (add_newlines) {
            for (int i = 0; i < 5; i++) {
                fprintf(log_f, "\n");
            }
        }
        fclose(log_f);
    }
}

// Function to get current timestamp
void get_timestamp(char *buffer, size_t size) {
    time_t now = time(NULL);
    struct tm *t = localtime(&now);
    strftime(buffer, size, "%Y-%m-%d %H:%M:%S", t);
}

// Function to send NEWNYM signal to Tor control port
int send_signal(int port, const char *password, char *status_buffer) {
    int sockfd;
    struct sockaddr_in serv_addr;
    char buffer[BUFFER_SIZE];
    int bytes_received;

    log_message("[TOR] Sending NEWNYM signal...", 0, 0);

    // Create socket
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        log_message("[ERROR] Socket creation failed", 0, 0);
        return 0;
    }

    // Set up server address
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(port);
    serv_addr.sin_addr.s_addr = inet_addr("127.0.0.1");

    // Connect to Tor control port
    if (connect(sockfd, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0) {
        log_message("[ERROR] Connection to Tor control port failed", 0, 0);
        close(sockfd);
        return 0;
    }

    // Authenticate with Tor control port
    if (password != NULL) {
        snprintf(buffer, sizeof(buffer), "AUTHENTICATE \"%s\"\r\n", password);
        send(sockfd, buffer, strlen(buffer), 0);
        bytes_received = recv(sockfd, buffer, sizeof(buffer), 0);
        if (bytes_received < 1 || strstr(buffer, "250") == NULL) {
            log_message("[ERROR] Tor authentication failed", 0, 0);
            close(sockfd);
            return 0;
        }
    }

    // Send NEWNYM signal
    send(sockfd, "SIGNAL NEWNYM\r\n", 15, 0);
    bytes_received = recv(sockfd, buffer, sizeof(buffer), 0);
    if (bytes_received < 1 || strstr(buffer, "250") == NULL) {
        log_message("[ERROR] Failed to send NEWNYM signal", 0, 0);
        close(sockfd);
        return 0;
    }

    // Get circuit status
    send(sockfd, "GETINFO circuit-status\r\n", 24, 0);
    bytes_received = recv(sockfd, status_buffer, BUFFER_SIZE - 1, 0);
    if (bytes_received > 0) {
        status_buffer[bytes_received] = '\0';
    } else {
        status_buffer[0] = '\0';
    }

    // Log the current circuit status
    log_message("[TOR] Current Circuit Status:", 0, 0);
    log_message(status_buffer, 0, 0);  // Log the received circuit status

    log_message("[TOR] New Tor Circuits Requested Successfully.", 0, 0);
    close(sockfd);
    return 1;
}

// Function to check if circuits are different
int circuits_are_different(const char *old_status, const char *new_status) {
    return strcmp(old_status, new_status) != 0;
}

int main() {
    char old_status_9051[BUFFER_SIZE] = {0};
    char old_status_9054[BUFFER_SIZE] = {0};
    char new_status_9051[BUFFER_SIZE] = {0};
    char new_status_9054[BUFFER_SIZE] = {0};

    const char *VANGUARD = getenv("VANGUARD");
    if (!VANGUARD) {
        log_message("Error: Tor control port password (VANGUARD) is not set.", 0, 1);
        return 1;
    }

    log_message("Starting Tor circuit refresh...", 0, 1);

    // Get initial circuit status
    if (!send_signal(TOR_CONTROL_PORT_1, VANGUARD, old_status_9051) ||
        !send_signal(TOR_CONTROL_PORT_2, VANGUARD, old_status_9054)) {
        return 1;
    }

    // Send NEWNYM signal again and compare statuses
    send_signal(TOR_CONTROL_PORT_1, VANGUARD, new_status_9051);
    send_signal(TOR_CONTROL_PORT_2, VANGUARD, new_status_9054);

    if (circuits_are_different(old_status_9051, new_status_9051) ||
        circuits_are_different(old_status_9054, new_status_9054)) {
        log_message("Tor circuits have been successfully refreshed.", 1, 1);
    } else {
        log_message("Tor circuits did not change. Retrying...", 0, 0);
    }

    log_message("Tor circuit refresh completed.", 1, 1);
    return 0;
}

File: Src/torflux/torflux.go
// Copyright(C) 2025 NOXCIS [https://github.com/NOXCIS]
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"bufio"
	"bytes"
	"flag"
	"fmt"
	"net"
	"os"
	"os/exec"
	"strings"
	"time"
)

const (
	logDir          = "./log"
	logFile         = logDir + "/tor_circuit_refresh_log.txt"
	bufferSize      = 8192
	torControlPort1 = 9051
	torControlPort2 = 9054
	socketTimeout   = 5 * time.Second
)

// logMessage logs messages to a file and optionally prints to the console.
func logMessage(message string, addNewlines, toConsole bool) {
	timestamp := time.Now().Format("2006-01-02 15:04:05")
	logMsg := fmt.Sprintf("[%s] %s", timestamp, message)

	if toConsole {
		fmt.Println(logMsg)
	}

	// Ensure log directory exists
	if err := os.MkdirAll(logDir, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "[ERROR] Could not create log directory: %v\n", err)
		return
	}

	// Open or create log file
	f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[ERROR] Could not open log file: %v\n", err)
		return
	}
	defer f.Close()

	writer := bufio.NewWriter(f)
	writer.WriteString(logMsg + "\n")
	if addNewlines {
		writer.WriteString(strings.Repeat("\n", 5))
	}
	writer.Flush()
}

// sendSignal sends a NEWNYM signal to the Tor control port and retrieves circuit status.
func sendSignal(port int, password string, statusBuffer *bytes.Buffer) bool {
	address := fmt.Sprintf("127.0.0.1:%d", port)
	conn, err := net.DialTimeout("tcp", address, socketTimeout)
	if err != nil {
		logMessage("[TOR-FLUX] [ERROR] Connection to Tor control port failed", false, false)
		return false
	}
	defer conn.Close()

	conn.SetDeadline(time.Now().Add(socketTimeout))

	reader := bufio.NewReader(conn)

	// Authenticate
	if password != "" {
		authCmd := fmt.Sprintf("AUTHENTICATE \"%s\"\r\n", password)
		fmt.Fprint(conn, authCmd)

		resp, _ := reader.ReadString('\n')
		if !strings.Contains(resp, "250") {
			logMessage("[TOR-FLUX] [ERROR] Tor authentication failed", false, false)
			return false
		}
	}

	// Send NEWNYM signal
	fmt.Fprint(conn, "SIGNAL NEWNYM\r\n")
	resp, _ := reader.ReadString('\n')
	if !strings.Contains(resp, "250") {
		logMessage("[TOR-FLUX] [ERROR] Failed to send NEWNYM signal", false, false)
		return false
	}

	// Get circuit status
	fmt.Fprint(conn, "GETINFO circuit-status\r\n")
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		statusBuffer.WriteString(line)
	}

	logMessage("[TOR-FLUX] Current Circuit Status:", false, false)
	logMessage(statusBuffer.String(), false, false)
	logMessage("[TOR-FLUX] New Tor Circuits Requested Successfully.", false, false)

	return true
}

// sendHUP sends a HUP signal to the specified Tor control port
func sendHUP(port int, password string) bool {
	address := fmt.Sprintf("127.0.0.1:%d", port)
	conn, err := net.DialTimeout("tcp", address, socketTimeout)
	if err != nil {
		logMessage("[TOR-FLUX] [ERROR] Connection to Tor control port failed", false, false)
		return false
	}
	defer conn.Close()

	conn.SetDeadline(time.Now().Add(socketTimeout))
	reader := bufio.NewReader(conn)

	// Authenticate
	if password != "" {
		authCmd := fmt.Sprintf("AUTHENTICATE \"%s\"\r\n", password)
		fmt.Fprint(conn, authCmd)

		resp, _ := reader.ReadString('\n')
		if !strings.Contains(resp, "250") {
			logMessage("[TOR-FLUX] [ERROR] Tor authentication failed", false, false)
			return false
		}
	}

	// Send SIGNAL HUP
	fmt.Fprint(conn, "SIGNAL HUP\r\n")
	resp, _ := reader.ReadString('\n')
	if !strings.Contains(resp, "250") {
		logMessage("[TOR-FLUX] [ERROR] Failed to send HUP signal", false, false)
		return false
	}

	logMessage("[TOR-FLUX] HUP signal sent successfully", false, true)
	return true
}

func startTor(configPath string) bool {
	cmd := exec.Command("tor", "-f", configPath)
	err := cmd.Start()
	if err != nil {
		logMessage(fmt.Sprintf("[TOR-FLUX] [ERROR] Failed to start Tor with config %s: %v", configPath, err), false, true)
		return false
	}
	
	logMessage(fmt.Sprintf("[TOR-FLUX] Started Tor with config %s", configPath), false, true)
	return true
}

func stopTor(port int, password string) bool {
	address := fmt.Sprintf("127.0.0.1:%d", port)
	conn, err := net.DialTimeout("tcp", address, socketTimeout)
	if err != nil {
		logMessage(fmt.Sprintf("[TOR-FLUX] [ERROR] Connection to Tor control port %d failed: %v", port, err), false, true)
		return false
	}
	defer conn.Close()

	conn.SetDeadline(time.Now().Add(socketTimeout))
	reader := bufio.NewReader(conn)

	// Authenticate
	if password != "" {
		authCmd := fmt.Sprintf("AUTHENTICATE \"%s\"\r\n", password)
		fmt.Fprint(conn, authCmd)

		resp, _ := reader.ReadString('\n')
		if !strings.Contains(resp, "250") {
			logMessage(fmt.Sprintf("[TOR-FLUX] [ERROR] Tor authentication failed on port %d", port), false, true)
			return false
		}
	}

	// Send SIGNAL HALT
	fmt.Fprint(conn, "SIGNAL HALT\r\n")
	resp, _ := reader.ReadString('\n')
	if !strings.Contains(resp, "250") {
		logMessage(fmt.Sprintf("[TOR-FLUX] [ERROR] Failed to send HALT signal to port %d", port), false, true)
		return false
	}

	logMessage(fmt.Sprintf("[TOR-FLUX] Tor process on port %d stopped successfully", port), false, true)
	return true
}

func main() {
	// Define command line flags
	configType := flag.String("config", "", "Configuration type (main/dns) for HUP signal")
	action := flag.String("action", "", "Action to perform (start/stop/hup/newnym)")
	flag.Parse()

	password := os.Getenv("VANGUARD")
	if password == "" {
		logMessage("[TOR-FLUX] [ERROR] Tor control port password (VANGUARD) is not set or empty.", false, true)
		os.Exit(1)
	}

	// Handle different actions
	if *action != "" {
		var port int
		var configPath string
		
		switch strings.ToLower(*configType) {
		case "main":
			port = torControlPort1
			configPath = "/etc/tor/torrc"
		case "dns":
			port = torControlPort2
			configPath = "/etc/tor/dnstorrc"
		default:
			logMessage("[TOR-FLUX] [ERROR] Invalid config type. Use 'main' or 'dns'.", false, true)
			os.Exit(1)
		}

		switch strings.ToLower(*action) {
		case "start":
			if !startTor(configPath) {
				os.Exit(1)
			}
			return
		case "stop":
			if !stopTor(port, password) {
				os.Exit(1)
			}
			return
		case "hup":
			if !sendHUP(port, password) {
				os.Exit(1)
			}
			return
		case "newnym":
			statusBuffer := &bytes.Buffer{}
			if !sendSignal(port, password, statusBuffer) {
				os.Exit(1)
			}
			return
		default:
			logMessage("[TOR-FLUX] [ERROR] Invalid action. Use 'start', 'stop', 'hup', or 'newnym'.", false, true)
			os.Exit(1)
		}
	}

	// Original NEWNYM functionality (default if no action specified)
	logMessage("[TOR-FLUX] Starting Tor circuit refresh...", false, true)
	statusBuffer1 := &bytes.Buffer{}
	statusBuffer2 := &bytes.Buffer{}
	sendSignal(torControlPort1, password, statusBuffer1)
	sendSignal(torControlPort2, password, statusBuffer2)
	logMessage("[TOR-FLUX] Tor circuit refresh completed.", true, true)
}
